File: ocaml/interp/value.ml
open Luaparser.Ast

type tkey =
  | KInt of Int64.t
  | KString of string

type t =
  | Nil
  | Bool of bool
  | Int of Int64.t
  | Float of float
  | String of string
  | Function of func
  | Table of (tkey, t) Hashtbl.t

and func =
  | Closure of name list * env * block
  | Print

and env = {
  globals : (name, t) Hashtbl.t;
  locals : (name, t) Hashtbl.t list;
}

let to_string = function
  | Nil -> "nil"
  | Bool b -> string_of_bool b
  | Int n -> Int64.to_string n
  | Float f ->
    let b = Buffer.create 15 in
    Printf.bprintf b "%.14g" f;
    if String.for_all (fun c -> c = '-' || '0' <= c && c <= '9') (Buffer.contents b) then
      Buffer.add_string b ".0";
    Buffer.contents b
  | String s -> s
  | Function _ -> "<function>"
  | Table _ -> "<table>"

let as_float = function
  | Float f -> f
  | Int n -> Int64.to_float n
  | v -> failwith (to_string v ^ ": not a float")

let as_bool = function
  | Bool b -> b
  | Nil -> false
  | _ -> true

let as_function = function
  | Function f -> f
  | v -> failwith (to_string v ^ ": not a function")

let as_table_key = function
  | Int n -> KInt n
  | String s -> KString s
  | v -> failwith (to_string v ^ ": not a table key (int or string)")

let as_table = function
  | Table t -> t
  | v -> failwith (to_string v ^ ": not a table")

let add v1 v2 =
  match v1, v2 with
  | Int(n1), Int(n2) -> Int(Int64.add n1 n2)
  | _, _ -> Float(as_float v1 +. as_float v2)

let sub v1 v2 =
  match v1, v2 with
  | Int(n1), Int(n2) -> Int(Int64.sub n1 n2)
  | _, _ -> Float(as_float v1 -. as_float v2)

let neg = function
  | Int(n) -> Int(Int64.neg n)
  | Float(f) -> Float(-.f)
  | v -> failwith (to_string v ^ ": not a number")

let mul v1 v2 =
  match v1, v2 with
  | Int(n1), Int(n2) -> Int(Int64.mul n1 n2)
  | _, _ -> Float(as_float v1 *. as_float v2)

let cmp_int_float n f =
  if Float.is_nan f then None
  else
    let c = compare (Int64.to_float n) f in
    if c < 0 then Some(-1)
    else if c > 0 then Some(1)
    else if n < Int64.of_float (ceil f) then Some(-1)
    else if n > Int64.of_float (floor f) then Some(1)
    else Some(0)

let equal v1 v2 =
  v1 == v2 ||
  match v1, v2 with
  | Nil, Nil -> true
  | Bool(b1), Bool(b2) -> b1 = b2
  | Int(n1), Int(n2) -> n1 = n2
  | Float(f), Int(n) | Int(n), Float(f) ->
    cmp_int_float n f = Some(0)
  | Float(f1), Float(f2) -> f1 = f2
  | String(s1), String(s2) -> s1 = s2
  | _, _ -> false

let lt v1 v2 =
  match v1, v2 with
  | Int(n1), Int(n2) -> n1 < n2
  | Float(f1), Int(n2) ->
    cmp_int_float n2 f1 = Some(1)
  | Int(n1), Float(f2) ->
    cmp_int_float n1 f2 = Some(-1)
  | Float(f1), Float(f2) -> f1 < f2
  | _, _ ->
    failwith ("trying to compare " ^ to_string v1 ^ " and " ^ to_string v2)

let le v1 v2 =
  match v1, v2 with
  | Int(n1), Int(n2) -> n1 <= n2
  | Float(f1), Int(n2) ->
    begin match cmp_int_float n2 f1 with Some(0|1) -> true | _ -> false end
  | Int(n1), Float(f2) ->
    begin match cmp_int_float n1 f2 with Some(0|(-1)) -> true | _ -> false end
  | Float(f1), Float(f2) -> f1 <= f2
  | _, _ ->
    failwith ("trying to compare " ^ to_string v1 ^ " and " ^ to_string v2)

let lookup_scope env name =
  try
    List.find (fun local_scope ->
      Hashtbl.mem local_scope name
    ) env.locals
  with Not_found ->
    env.globals

let lookup_ident env name =
  try Hashtbl.find (lookup_scope env name) name with
    Not_found -> Nil

let set_ident env name v =
  Hashtbl.replace (lookup_scope env name) name v

File: rust/src/interp/value.rs
use crate::{
    interp::env::LEnv,
    parser::ast::{Block, Name},
};
use std::{
    cell::RefCell,
    cmp::Ordering::{self, *},
    collections::HashMap,
    fmt::{Display, Formatter},
    ops::{Add, Mul, Neg, Sub},
    rc::Rc,
};

// Le type des valeurs de l'interpréteur.
// Celui-ci (ainsi que de nombreux autres types dans l'interpréteur) dépend
// d'un paramètre de lifetime 'ast représentant la durée de vie des emprunts de
// valeurs provenant de l'AST du programme. L'AST reste alloué et gelé pendant
// toute l'exécution de l'interpréteur, ce qui permet à cette durée de vie de
// rester vivante pendant toute l'exécution de l'interpréteur.
#[derive(Clone)]
pub enum Value<'ast> {
    Nil,
    Bool(bool),
    Number(Number),
    String(Rc<String>),
    Function(Function<'ast>),
    Table(Rc<RefCell<Table<'ast>>>),
}

// Type des nombres utilisées dans un programme mini-lua
#[derive(Debug, Clone, Copy)]
pub enum Number {
    Integer(i64),
    Float(f64),
}
use Number::*;

// Les fonctions sont soit la primitive "print", soit une fermeture,
// contenant une liste de noms de paramètres, un environnement local et une
// référence vers un bloc de l'AST constituant le corps de la fermeture.
#[derive(Clone)]
pub enum Function<'ast> {
    Print,
    Closure(&'ast [Name], Rc<LEnv<'ast>>, &'ast Block),
}

pub type Table<'ast> = HashMap<TKey, Value<'ast>>;

// Type des entrées de table de hachage
#[derive(Hash, PartialEq, Eq)]
pub enum TKey {
    KInt(i64),
    KString(Rc<String>),
}

// Fonctions privées sur le type Number. Elle ne sont utilisées que dans le
// module courant.
impl Add<Number> for Number {
    type Output = Number;
    fn add(self, rhs: Self) -> Number {
        match (self, rhs) {
            (Integer(n1), Integer(n2)) => Integer(n1 + n2),
            (Integer(n1), Float(f2)) => Float(n1 as f64 + f2),
            (Float(f1), Integer(n2)) => Float(f1 + n2 as f64),
            (Float(f1), Float(f2)) => Float(f1 + f2),
        }
    }
}

impl Sub<Number> for Number {
    type Output = Number;
    fn sub(self, rhs: Self) -> Number {
        match (self, rhs) {
            (Integer(n1), Integer(n2)) => Integer(n1 - n2),
            (Integer(n1), Float(f2)) => Float(n1 as f64 - f2),
            (Float(f1), Integer(n2)) => Float(f1 - n2 as f64),
            (Float(f1), Float(f2)) => Float(f1 - f2),
        }
    }
}

impl Neg for Number {
    type Output = Number;
    fn neg(self) -> Number {
        match self {
            Integer(n) => Integer(-n),
            Float(f) => Float(f),
        }
    }
}

impl Mul<Number> for Number {
    type Output = Number;
    fn mul(self, rhs: Self) -> Number {
        match (self, rhs) {
            (Integer(n1), Integer(n2)) => Integer(n1 * n2),
            (Integer(n1), Float(f2)) => Float(n1 as f64 * f2),
            (Float(f1), Integer(n2)) => Float(f1 * n2 as f64),
            (Float(f1), Float(f2)) => Float(f1 * f2),
        }
    }
}

impl PartialEq for Number {
    fn eq(&self, other: &Self) -> bool {
        self.partial_cmp(other) == Some(Equal)
    }
}

impl PartialOrd for Number {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        fn if_cmp(n: i64, f: f64) -> Option<Ordering> {
            Some((n as f64).partial_cmp(&f)?.then_with(|| {
                if n < f.ceil() as i64 {
                    Less
                } else if n > f.floor() as i64 {
                    Greater
                } else {
                    Equal
                }
            }))
        }
        match (*self, *other) {
            (Integer(n1), Integer(n2)) => n1.partial_cmp(&n2),
            (Integer(n1), Float(f2)) => if_cmp(n1, f2),
            (Float(f1), Integer(n2)) => if_cmp(n2, f1).map(Ordering::reverse),
            (Float(f1), Float(f2)) => f1.partial_cmp(&f2),
        }
    }
}

// Le trait Display permet d'afficher une valeur lors de l'exécution de
// l'instruction print. Dans le code de l'interpréteur, on peut ainsi utiliser
// l'instruction print!("{}", v) pour afficher la valeur v.
impl<'ast> Display for Value<'ast> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        use Value::*;
        match self {
            Nil => write!(f, "nil"),
            Bool(b) => write!(f, "{}", b),
            Number(Float(x)) => write!(f, "{:?}", x),
            Number(Integer(x)) => write!(f, "{}", x),
            String(str) => write!(f, "{}", str),
            Function(..) => write!(f, "<function>"),
            Table(..) => write!(f, "<table>"),
        }
    }
}

// Les fonctions suivantes peuvent être utilisées dans l'interpréteur pour
// implémenter les fonctionnalités correspondantes pour mini-lua.

impl<'ast> Value<'ast> {
    pub fn as_bool(&self) -> bool {
        !matches!(self, Value::Bool(false) | Value::Nil)
    }

    pub fn as_number(self) -> Number {
        if let Value::Number(res) = self {
            res
        } else {
            panic!("{} : not a number", self)
        }
    }

    pub fn as_function(self) -> Function<'ast> {
        if let Value::Function(res) = self {
            res
        } else {
            panic!("{} : not a function", self)
        }
    }

    pub fn as_table(self) -> Rc<RefCell<Table<'ast>>> {
        if let Value::Table(res) = self {
            res
        } else {
            panic!("{} : not a table", self)
        }
    }

    pub fn as_table_key(self) -> TKey {
        match self {
            Value::Number(Number::Integer(n)) => TKey::KInt(n),
            Value::String(s) => TKey::KString(s),
            _ => panic!("{} : not a table key (int or string)", self),
        }
    }

    pub fn add(self, other: Self) -> Self {
        Value::Number(self.as_number() + other.as_number())
    }

    pub fn sub(self, other: Self) -> Self {
        Value::Number(self.as_number() - other.as_number())
    }

    pub fn mul(self, other: Self) -> Self {
        Value::Number(self.as_number() * other.as_number())
    }

    pub fn neg(self) -> Self {
        Value::Number(-self.as_number())
    }

    pub fn lt(self, other: Self) -> bool {
        match (self, other) {
            (Value::Number(n1), Value::Number(n2)) => n1 < n2,
            (s, o) => panic!("trying to compare {} and {}", s, o),
        }
    }

    pub fn le(self, other: Self) -> bool {
        match (self, other) {
            (Value::Number(n1), Value::Number(n2)) => n1 <= n2,
            (s, o) => panic!("trying to compare {} and {}", s, o),
        }
    }
}

// L'opérateur == peut être utiliser pour tester l'égalité de deux valeurs.
impl<'ast> PartialEq for Value<'ast> {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Value::Bool(b1), Value::Bool(b2)) => b1 == b2,
            (Value::Nil, Value::Nil) => true,
            (Value::Number(n1), Value::Number(n2)) => n1 == n2,
            (Value::String(s1), Value::String(s2)) => s1 == s2,
            (Value::Function(f1), Value::Function(f2)) => std::ptr::eq(f1, f2),
            (Value::Table(t1), Value::Table(t2)) => std::ptr::eq(t1, t2),
            _ => false,
        }
    }
}
